File: laravelBackend1/app/Services/UserAuthService.php
<?php

namespace App\Services;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Http;


class UserAuthService {


    public function authenticateUser($userId, Request $request) {
        try {
            $authTokenVal = $request->cookie("authToken$userId");
            $refreshTokenVal = $request->cookie("refreshToken$userId");

            $authTokenIsValidlyStructured = true;
            try {
                $decodedTokenBytes = base64_decode($authTokenVal, true);
                if ($decodedTokenBytes === false || strlen($decodedTokenBytes) != 100) {
                    $authTokenIsValidlyStructured = false;
                }
            } catch (\Exception $e) {
                $authTokenIsValidlyStructured = false;
            }

            if (!$authTokenIsValidlyStructured) {
                return 'The provided authUser token, if any, in your cookies has an invalid structure.';
            }

            try {
                $decodedTokenBytes = base64_decode($refreshTokenVal, true);
                if ($decodedTokenBytes === false || strlen($decodedTokenBytes) != 100) {
                    $refreshTokenVal = "";
                }
            } catch (\Exception $e) {
                $refreshTokenVal = "";
            }

            $cookiesText = "authToken$userId=$authTokenVal;";
            if (strlen($refreshTokenVal) > 0) {
                $cookiesText .= " refreshToken$userId=$refreshTokenVal;";
            }

            $response = Http::withHeaders([
                'Cookie' => $cookiesText
            ])->get("http://34.111.89.101/api/Home-Page/expressJSBackend1/authenticateUser/$userId");

            if ($response->failed()) {
                return false;
            }

            $setCookies = $response->header('Set-Cookie');
            if ($setCookies) {
                $setCookiesArray = is_array($setCookies) ? $setCookies : [$setCookies];

                foreach ($setCookiesArray as $cookie) {
                    $cookieKeyAndValue = explode(';', $cookie)[0];
                    $cookieParts = explode('=', $cookieKeyAndValue);

                    if (trim($cookieParts[0]) === "authToken{$userId}") {
                        $refreshedAuthToken = trim($cookieParts[1]);

                        foreach (explode(';', $cookie) as $attribute) {
                            $trimmedAttribute = trim($attribute);
                            if (str_starts_with(strtolower($trimmedAttribute), 'expires=')) {
                                $expiresValue = substr($trimmedAttribute, 8);
                                try {
                                    $refreshedAuthTokenCookieExpiration = new \DateTime($expiresValue);
                                    return [
                                        $refreshedAuthToken,
                                        $refreshedAuthTokenCookieExpiration->format('Y-m-d H:i:s')
                                    ];
                                }
                                catch (\Exception $e) {
                                    break;
                                }
                            }
                        }
                        break;
                    }
                }
            }

            return true;
        }
        catch (\Exception $e) {
            return 'There was trouble connecting to the ExpressJS backend for user-authentication';
        }
    }
}

File: pythonWebSocket/services/UserAuthService.py
import requests
import base64
from datetime import datetime


class UserAuthService:


    def __init__(self):
        pass


    def authenticate_user(cookies, user_id):
        try:
            auth_token_val = cookies.get(f'authToken{user_id}')
            refresh_token_val = cookies.get(f'refreshToken{user_id}')

            auth_token_is_validly_structured = True
            try:
                decoded_token_bytes = base64.b64decode(auth_token_val, validate=True)
                if not decoded_token_bytes or len(decoded_token_bytes) != 100:
                    auth_token_is_validly_structured = False
            except:
                auth_token_is_validly_structured = False

            if not auth_token_is_validly_structured:
                return 'The provided authUser token, if any, in your cookies has an invalid structure.'

            try:
                decoded_token_bytes = base64.b64decode(refresh_token_val, validate=True)
                if not decoded_token_bytes or len(decoded_token_bytes) != 100:
                    refresh_token_val = ''
            except:
                refresh_token_val = ''

            cookies_text = f'authToken{user_id}={auth_token_val};'
            if refresh_token_val:
                cookies_text += f' refreshToken{user_id}={refresh_token_val};'

            response = requests.get(
                f'http://34.111.89.101/api/Home-Page/expressJSBackend1/authenticateUser/{user_id}',
                headers={'Cookie': cookies_text},
            )

            if not response.ok:
                return False

            set_cookies = response.headers.get('Set-Cookie')
            if set_cookies:
                set_cookies_array = set_cookies if isinstance(set_cookies, list) else [set_cookies]

                for cookie in set_cookies_array:
                    cookie_key_and_value = cookie.split(';')[0]
                    cookie_parts = cookie_key_and_value.split('=')

                    if cookie_parts[0].strip() == f'authToken{user_id}':
                        refreshed_auth_token = cookie_parts[1].strip()

                        for attribute in cookie.split(';'):
                            trimmed_attribute = attribute.strip()
                            if trimmed_attribute.lower().startswith('expires='):
                                expires_value = trimmed_attribute[8:]
                                try:
                                    refreshed_auth_token_cookie_expiration = datetime.strptime(
                                        expires_value, '%a, %d-%b-%Y %H:%M:%S GMT'
                                    )
                                    return [
                                        refreshed_auth_token,
                                        refreshed_auth_token_cookie_expiration.strftime('%Y-%m-%d %H:%M:%S'),
                                    ]
                                except:
                                    break
                        break

            return True
        except:
            return 'There was trouble connecting to the ExpressJS backend for user authentication'
